### ISSUE TITLE
Index set concatenation, please?

### ISSUE THREAD

**Issue by axch** [google-research/dex-lang#716]
I tried this in a notebook:
```
for i : (Fin 5 | Fin 3). i
```

This caused the `dex web` server to repeatedly say
```
Not implemented (Fin 5 | Fin 3)
CallStack (from HasCallStack):
  error, called at src/lib/Builder.hs:948:20 in dex-0.1.0.0-KC2OgnYnAXsK6o9fCGTvOZ:Builder
["getnext"]
```

I get about 20 of those, with a pause between each (maybe one per notebook cell?).

What I was actually trying to do was stitch a timeseries (represented as (Fin N) => Float) together with a forecast of it (represented as (Fin M) => Float) to plot both together, but that didn't seem to work.

**Comment by axch** [comment]
For the record, I can accomplish my proximate goal with this lovely:
```
def array_concat (xs1: n1=>a) (xs2: n2=>a) : n3=>a =
  sz = size n1
  for i. if (ordinal i) < sz
    then xs1.(fromOrdinal n1 $ ordinal i)
    else xs2.(fromOrdinal n2 $ (ordinal i) - sz)

concatted : (Fin 409) = array_concat series forecast_means
```
but sadly, I now have to count the size of my data set myself.

**Comment by tscholak** [comment]
oh, so you want `(Fin 5 | Fin 3) ~ Fin 8`?

btw,
```
for i:(Fin 5 & Fin 3). I
```
gives
```
[ ((0@Fin 5), (0@Fin 3))
, ((0@Fin 5), (1@Fin 3))
, ((0@Fin 5), (2@Fin 3))
, ((1@Fin 5), (0@Fin 3))
, ((1@Fin 5), (1@Fin 3))
, ((1@Fin 5), (0@Fin 3))
, ((1@Fin 5), (1@Fin 3))
, ((1@Fin 5), (2@Fin 3))
, ((2@Fin 5), (0@Fin 3))
, ((2@Fin 5), (1@Fin 3))
, ((2@Fin 5), (0@Fin 3))
, ((2@Fin 5), (1@Fin 3))
, ((2@Fin 5), (2@Fin 3))
, ((3@Fin 5), (0@Fin 3))
, ((3@Fin 5), (1@Fin 3)) ]@(Fin 5 & Fin 3)
```
for me. this looks wrong. shouldn't it be
```
[ ((0@Fin 5), (0@Fin 3))
, ((0@Fin 5), (1@Fin 3))
, ((0@Fin 5), (2@Fin 3))
, ((1@Fin 5), (0@Fin 3))
, ((1@Fin 5), (1@Fin 3))
, ((1@Fin 5), (2@Fin 3))
, ((2@Fin 5), (0@Fin 3))
, ((2@Fin 5), (1@Fin 3))
, ((2@Fin 5), (2@Fin 3))
, ((3@Fin 5), (0@Fin 3))
, ((3@Fin 5), (1@Fin 3))
, ((3@Fin 5), (2@Fin 3))
, ((4@Fin 5), (0@Fin 3))
, ((4@Fin 5), (1@Fin 3))
, ((4@Fin 5), (2@Fin 3)) ]@(Fin 5 & Fin 3)
```
?
I'm on `safe-names-dev`, commit https://github.com/google-research/dex-lang/commit/74c1a61537adee94339ca82b1ebb418bc7281915

**Comment by dougalm** [comment]
It's another duplicate of #186, I'm afraid. `(|)` is just a prelude-defined ADT (just a cute renaming of Haskell's `Either`) and we don't have user-defined index sets. (Adam's working on it though.) The workaround in the meantime has been to use variant types rather than `(|)` since these are built-in and we internally know how to treat them as index sets. So it's:

```
for i : { a: Fin 5 | b: Fin 3 }. i
> [ {| a = 0@Fin 5 |}
> , {| a = 1@Fin 5 |}
> , {| a = 2@Fin 5 |}
> , {| a = 3@Fin 5 |}
> , {| a = 4@Fin 5 |}
> , {| b = 0@Fin 3 |}
> , {| b = 1@Fin 3 |}
> , {| b = 2@Fin 3 |} ]@{a: Fin 5 | b: Fin 3}
```

**Comment by dougalm** [comment]
@tscholak , thanks for noticing the bug in `for i:(Fin 5 & Fin 3). i` on safe-names-dev . It works fine on main so I must have broken something. I'll add it as a test if it's not already being caught elsewhere.

**Comment by axch** [comment]
Dougal's comment solves my proximate problem much better than the previously posted workaround, but I'll leave this open to track the `for i:(Fin 5 & Fin 3). i` issue.

**Comment by dougalm** [comment]
Update on the `for i:(Fin 5 & Fin 3). i` bug: I missed a `reverse` in int-to-index conversion for product types. It only shows up when you print literal indexes because there's (unfortunately) a separate path for that. Fixed by 13b35df.
